File: api/report.py
"""경영진 보고서 작성기 API."""

import json
from typing import Literal, Optional

from fastapi import APIRouter
from fastapi.concurrency import run_in_threadpool
from fastapi.responses import Response, StreamingResponse
from pydantic import BaseModel, Field

from services import report as rp

router = APIRouter(prefix="/api/report", tags=["report"])


@router.get("/options")
def options():
    return {"styles": rp.REPORT_STYLES, "depths": rp.REPORT_DEPTHS, "modes": rp.RESEARCH_MODES,
            "style_help": rp.STYLE_HELP, "depth_help": rp.DEPTH_HELP}


class ReportRequest(BaseModel):
    topic: str = Field(min_length=1, max_length=6000)
    style: Literal["CEO/임원 보고", "전략 검토", "이슈·리스크 보고", "시장·경쟁 분석", "규제·법률 검토", "사업/투자 검토"] = "CEO/임원 보고"
    depth: Literal["핵심 중심", "표준", "상세"] = "표준"
    mode: Literal["빠른 작성", "최신자료 포함"] = "빠른 작성"


@router.post("/generate")
def generate(req: ReportRequest):
    """SSE: event tool(도구 호출 진행) → answer(보고서) 또는 error → done"""
    def stream():
        for event in rp.run_report(req.topic, req.style, req.depth, req.mode):
            kind = event.pop("type")
            yield f"event: {kind}\ndata: {json.dumps(event, ensure_ascii=False)}\n\n"
        yield "event: done\ndata: {}\n\n"

    return StreamingResponse(stream(), media_type="text/event-stream",
                             headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"})


class Chart(BaseModel):
    company: Optional[str] = None
    dates: list[str] = Field(default_factory=list, max_length=500)


class PdfRequest(BaseModel):
    text: str = Field(min_length=1, max_length=40000)
    chart: Optional[Chart] = None


@router.post("/pdf")
async def pdf(req: PdfRequest):
    data = await run_in_threadpool(rp.report_pdf, req.text, req.chart.model_dump() if req.chart else None)
    return Response(data, media_type="application/pdf",
                    headers={"Content-Disposition": 'attachment; filename="executive_report.pdf"'})

File: web/src/lib/report.ts
// FastAPI /api/report 클라이언트
import { apiFetch } from "@/lib/access";
import { postSSE } from "./sse";

const API_URL = (process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000").replace(/\/$/, "");

export type ReportOptions = {
  styles: string[];
  depths: string[];
  modes: string[];
  style_help: Record<string, string>;
  depth_help: Record<string, string>;
};

export type ReportChart = { company: string; labels: string[]; values: number[]; dates: string[] } | null;
export type ToolStep = { name: string; label: string; input: Record<string, unknown> };
export type ReportResult = { text: string; chart: ReportChart };

export const reportApi = {
  options: async (): Promise<ReportOptions> => {
    const res = await apiFetch(`${API_URL}/api/report/options`);
    if (!res.ok) throw new Error(`서버 응답 오류 (${res.status})`);
    return res.json();
  },

  generate: async (
    body: { topic: string; style: string; depth: string; mode: string },
    onTool: (t: ToolStep) => void,
    signal?: AbortSignal,
  ): Promise<ReportResult> => {
    let result: ReportResult | null = null;
    let error = "";
    await postSSE(
      `${API_URL}/api/report/generate`,
      body,
      (event, data) => {
        if (event === "tool") onTool(data as ToolStep);
        else if (event === "answer") result = data as ReportResult;
        else if (event === "error") error = (data as { message: string }).message;
      },
      signal,
    );
    if (error) throw new Error(error);
    if (!result) throw new Error("보고서를 받지 못했어요.");
    return result;
  },

  pdf: async (text: string, chart: ReportChart): Promise<Blob> => {
    const res = await apiFetch(`${API_URL}/api/report/pdf`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ text, chart: chart ? { company: chart.company, dates: chart.dates } : null }),
    });
    if (!res.ok) throw new Error(`PDF 생성 실패 (${res.status})`);
    return res.blob();
  },
};
